### DISCUSSION TITLE
Star - Constellation Quiz Scripts - Test your Knowledge

### DISCUSSION THREAD

**Discussion by qam1** [Stellarium/stellarium#2662] [Show and tell]
Some scripts I wrote that are not quite ready, but I see in the coming new version of Stellarium the scripting engine may not work properly and Stellarium won't work unless you have Windows 10 or above, so I might as well put these out there now since they are fun.

These scripts will test how well you know your star names. The scripts are a quiz/game on how well you know your star and Constellation names.

The Quizzes display a star/constellation and multiple choices which you answer using your keyboard (Sorry they currently only designed for the American English keyboards)  

3 Scripts, 
1) Quiz on the 60 Brightest stars (I would have made 50 but Stellarium has 8 stars tied at 50, so I went with 60 to get 'em all).  
2) Quiz on Prominent stars visible in the Northern Hemisphere
3) Quiz on the Constellations  

The 2 Star Quizzes could use a little more fine tuning but seem to work OK, the Constellation script on the other hand has potential but unfortunately is a bit wonky and should be considered a Beta version only.   

_They can be downloaded here._

http://www.alienbasecamp.com/Stellarium/quizzes.zip

I didn't write an instruction manual/FAQ yet, but they should be easy enough to follow. Just download the file and extract in the script folder. Remember use the keyboard letters to select game and answer questions. No mouse or touch screen support.

**Comment by gzotti** [comment]
Do I understand they use the same keyboard hacks which no longer work with Qt6 scripting? We can host them as optional downloads, but must mark them as "Qt5/version 0.* only".

**Reply by qam1** [reply]
I don't know right now, I have limited access to a Windows 10 computer, but I would guess no.

These script on the beta the keypresses works when you choose the quiz, but then crashes when the quiz starts. There is no real difference between the keypress "commands" at the start and later parts of the script, so I don't think that's what is causing the crash. Also other scripts of mine like earth1.ssc that comes with Stellarium that use keypress "commands" work fine and don't crash. But maybe this weekend I will be able to use a Windows 10 computer again and hopefully figure out what the issue is.

There is an issue with scripts using the date keys, but there is an easy fix

**Comment by Dumbo-programmer** [comment]
hey man will you ever continue or develop it

**Comment by kutaibaa-akraa** [comment]
Hi @qam1 

First of all - **thank you  so much @qam1** for sharing this Star Quiz script. I had a blast playing it, and it's genuinely useful for learning star names. 

I was so inspired by your work that I took your "Star Quiz - Northern Hemisphere Eddition" script and, with the help of AI, created a significantly **optimized and restructured version 2.0**. I wanted to share the key improvements here and ask for your permission to use/adapt this enhanced version in my own Stellarium projects.

Here's what was fundamentally changed:

### Critical Fixes:
1. **Array bounds bug** - Original loop `for(i=0; i<n+1; i++)` accessed index 116 which doesn't exist. Fixed to proper bounds.
2. **Array initialization** - `new Array(length= n)` is invalid JS syntax. Corrected to `new Array(n+1)`.
3. **Fisher-Yates shuffle** - Replaced the `for(i=0; i<1200; i++)` fixed-iteration shuffle with a proper O(n) shuffle algorithm.
4. **Duplicate prevention** - Added `starUsageCount` system to prevent the correct answer star from appearing among wrong options (which was possible in original).

### Performance & Architecture:
5. **Star info caching** - All `getObjectInfo()` calls are done once upfront into `starInfoCache{}` instead of calling repeatedly during gameplay. Dramatic speed improvement visible with 50 questions.
6. **Object-oriented design** - Wrapped game state into a `StarQuizGame` class for cleaner state management.
7. **`core.debug()` logging** - Comprehensive debug output visible right in Stellarium's script console window (much more useful than `console.log`).
8. **Proper error handling** - Added `try-catch` in `main()` that gracefully restores Stellarium state on any crash.

### Smart Features Added:
9. **Physics-aware wrong answers** - All 4 wrong options are now selected as the *closest angular-distance stars* to the correct one (not just 1 close + 3 random).
10. **Fixed Schedar/Shedar detection logic** - The original condition was inverted.
11. **Graded performance feedback** - 5 different achievement messages based on score percentage.

Here's proof it works - my first playthrough debug log showing 86% (43/50): 

```code
Starting script at Sun Jun 28 21:58:53 2026
========================================
Star Quiz v2.0 - Game Starting
========================================
Fixed star name: Shedar -> Schedar
Caching star information...
Cached 115 stars successfully
Difficulty selected: 50 questions
Game initialized with 50 questions
Star pool size: 115 stars available
Warning: No cached info for correct star Mintaka
Question 1/50
Correct star: Mintaka
Options: Sabik, Gomeisa, Sargas, Alcyone, Mintaka
Best viewing date set for Mintaka (Altitude: 44.73°)
Answer detected: 5
CORRECT!
Warning: No cached info for correct star Sheliak
Question 2/50
Correct star: Sheliak
Options: Sheliak, Hatysa, Diphda, Unukalhai, Zubeneschamali
Best viewing date set for Sheliak (Altitude: 77.36°)
Answer detected: 1
CORRECT!
Warning: No cached info for correct star Elnath
Question 3/50
Correct star: Elnath
Options: Elnath, Larawag, Vindemiatrix, Tianguan, Izar
Best viewing date set for Elnath (Altitude: 73.63°)
Answer detected: 3
WRONG! Correct was: A)
.
.
.
Warning: No cached info for correct star Wolf 359
Question 50/50
Correct star: Wolf 359
Options: Spica, Wolf 359, Alkaid, Barnard's Star, Merak
Best viewing date set for Wolf 359 (Altitude: -999.00°)
Answer detected: 1
WRONG! Correct was: B)
Final Score: 43/50 (86%)
========================================
Game Over - Thank you for playing!
========================================
Script finished at Mon Jun 29 00:28:40 2026
```

I would be happy to use this enhanced version in my scripts, with full credit to you as the original author (your name remains in the header).

Thanks for sharing your work  
@Dumbo-programmer  see this 
Here is the new adapted script :  
[star_quiz_northern_hemisphere_edition_v20.zip](https://github.com/user-attachments/files/29444892/star_quiz_northern_hemisphere_edition_v20.zip)

<img width="942" height="649" alt="script_meta" src="https://github.com/user-attachments/assets/9031b598-2a84-4305-a587-19e860847386" />

---

**Reply by gzotti** [reply]
Very impressive!

**Comment by kutaibaa-akraa** [comment]
Hi @qam1,

Following up on the Star Quiz enhancement, I've applied the same optimization framework to your other two brilliant scripts. Here are the results:

### Constellations Quiz v2.0 
[constellations_quiz_v2.0.zip](https://github.com/user-attachments/files/29629600/constellations_quiz_v2.0.zip)


**Game-specific improvements:**

| Feature | Original | Enhanced |
|---------|----------|----------|
| Observer positioning | 3-tier if-else | 6-tier system with per-constellation declination metadata |
| Wrong answer selection | Random only | 2 nearby + 2 distant constellations for balanced challenge |
| Constellation isolation cleanup | Manual reset | Dedicated `cleanupConstellationDisplay()` between questions |
| Duplicate line removal | `setFlagArt(false)` called twice | Clean single calls |
| Key C functionality | Missing entirely | Restored as "Toggle Constellation Boundaries" option |

### Star Quiz - 60 Brightest Stars v2.0 
[star_quiz_60_brightest_stars_v2.0.zip](https://github.com/user-attachments/files/29629668/star_quiz_60_brightest_stars_v2.0.zip)

**Game-specific improvements:**

| Feature | Original | Enhanced |
|---------|----------|----------|
| Game mode filtering | 3 separate while-loops with duplicated logic | Single `applyGameModeFilter()` with mode constants |
| Filter verification | Silent failure if insufficient stars | Validates star count matches criteria, adjusts automatically |
| Star metadata | Fetched on-demand repeatedly | Pre-cached with ra/dec/vmag for all 60 stars |
| Wrong answer logic | Inconsistent (3 wrong + 1 close) | Consistent 4 wrong options (2 close + 2 distant) |
| Key F support | Present but fragile | Properly integrated with state restoration including `FlagFog` |

**Shared enhancements across both scripts (same as Star Quiz v2.0):**
- Fisher-Yates shuffle replacing 1200-iteration loops
- OOP architecture with dedicated game state classes
- `core.debug()` comprehensive logging throughout
- `try-catch` error recovery restoring Stellarium to usable state
- Array bounds fixes (`i < n+1` corrected)
- Star/constellation info caching for performance
- `MAX_REUSE_COUNT` preventing answer repetition

Both scripts tested and working smoothly on Stellarium 26.1.
